espresso/environment/3_3_fast_derivation_and_caff_node_test.go

```diff
@@ -0,0 +1,167 @@
+	}
+
+	// Shut down the Caff Node
+	defer env.Stop(t, caffNode)
+
+	addressAlice := system.Cfg.Secrets.Addresses().Alice
+	l1Client := system.NodeClient(e2esys.RoleL1)
+	l2Verif := system.NodeClient(e2esys.RoleVerif)
+	caffVerif := system.NodeClient(env.RoleCaffNode)
+
+	// We want to send some transactions from Bob to Alice
+	{
+		privateKey := system.Cfg.Secrets.Bob
+		bobOptions, err := bind.NewKeyedTransactorWithChainID(privateKey, system.Cfg.L1ChainIDBig())
+		if have, want := err, error(nil); have != want {
+			t.Fatalf("failed to create transaction options for bob:\nhave:\n\t\"%v\"\nwant:\n\t\"%v\"\n", have, want)
+		}
+
+		mintAmount := new(big.Int).SetUint64(1)
+		bobOptions.Value = mintAmount
+		_ = helpers.SendDepositTx(t, system.Cfg, l1Client, l2Verif, bobOptions, func(l2Opts *helpers.DepositTxOpts) {
+			// Send from Bob to Alice
+			l2Opts.ToAddr = addressAlice
+		})
+	}
+
+	// Initialize ticker to fire every 2 seconds
+	ticker := time.NewTicker(2 * time.Second)
+	defer ticker.Stop()
+
+	finishticker := time.NewTicker(30 * time.Second)
+	defer finishticker.Stop()
+
+	lastHead, err := l2Verif.BlockByNumber(ctx, nil)
+	if err != nil {
+		t.Fatalf("Failed to get initial l2Verif block: %v", err)
+	}
+
+	lastCaffHead, err := caffVerif.BlockByNumber(ctx, nil)
+	if err != nil {
+		t.Fatalf("Failed to get initial caffVerif block: %v", err)
+	}
+
+	// checkNewBlocks checks for new blocks and verifies their timestamps
+	checkNewBlocks := func(client *ethclient.Client, lastBlock *types.Block, nodeName string, logger string) (*types.Block, error) {
+		newBlock, err := client.BlockByNumber(ctx, nil)
+		if err != nil {
+			return nil, fmt.Errorf("Failed to get new %s block: %v", nodeName, err)
+		}
+
+		// Skip the warm-up period and check for new block by comparing the latest hash
+		if newBlock.Number().Cmp(big.NewInt(0)) != 0 && newBlock.Hash() == lastBlock.Hash() {
+			// we only report an error here, but not return a failure
+			log.Info("No new block for RPC after 2 seconds", "node", nodeName, "current number", newBlock.Number())
+		} else {
+			// instead, we check block timestamps
+			lastBlockTime := lastBlock.Time()
+			for j := new(big.Int).Add(lastBlock.Number(), big.NewInt(1)); j.Cmp(newBlock.Number()) <= 0; j.Add(j, big.NewInt(1)) {
+				block, err := client.BlockByNumber(ctx, j)
+				if err != nil {
+					return nil, fmt.Errorf("Failed to get block from %s: %v", nodeName, err)
+				}
+				// check the block timestamp
+				if have, want := block.Time(), lastBlockTime+uint64(1); have != want {
+					return nil, fmt.Errorf("Block timestamp mismatch:\nhave:\n\t\"%v\"\nwant:\n\t\"%v\"\n", have, want)
+				}
+				lastBlockTime = block.Time()
+			}
+		}
+		return newBlock, nil
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+			// Check for new block of op-node
+			newHead, err := checkNewBlocks(l2Verif, lastHead, "op-node", "l2verifier")
+			if err != nil {
+				t.Fatal(err)
```

Comment:
This `t.Fatal` would be more useful if it was a `t.Fatalf` with a string description that could uniquely identify it.  As it stands we cannot tell if the `err` returned here would come from `:150` or `:157` from the message alone.